### PR TITLE
Reduce number of defenders during offense play if enemy team has less robots

### LINKED ISSUE #2543: Adapt offense/defense to number of friendly and enemy robots on the field
### Description of the task

Subtask of #2440

### Acceptance criteria

- [ ] implement logic to adapt tactic assignment based on number of friendly and enemy robots on the field

### PR DESCRIPTION
**Pull request by xyztiger**
### Description

This PR changes our offense play to reduce the number of defenders if the enemy team has 3 or fewer robots.

### Resolved Issues

resolves #2543 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/ai/hl/stp/play/offense/offense_play.cpp [hunk 1]
```diff
@@ -26,20 +26,26 @@ void OffensePlay::getNextTactics(TacticCoroutine::push_type &yield, const World
...
     {
         return;
     }
-    if (play_update.num_tactics <= 3)
+
+    PriorityTacticVector tactics_to_return;
+    unsigned int num_defenders     = 2;
+    unsigned int num_enemy_robots  = static_cast<int>(play_update.world.enemyTeam().numRobots());
+
+    if (num_enemy_robots <= 3)
```
Comment (xyztiger) [inline src/software/ai/hl/stp/play/offense/offense_play.cpp:41]: the logic here is that they have at most half of a full team so we should need at most half of our normal number of defenders

File: src/software/ai/hl/stp/play/offense/offense_play.cpp [hunk 2]
```diff
@@ -26,20 +26,26 @@ void OffensePlay::getNextTactics(TacticCoroutine::push_type &yield, const World
...
+    {
+        num_defenders = num_enemy_robots > 1 ? 1 : 0;
```
Comment (xyztiger) [inline src/software/ai/hl/stp/play/offense/offense_play.cpp]: if they only have 1 robot we assume it's the goalie and we thus don't need any defenders

Comment (amrm3lm) [inline src/software/ai/hl/stp/play/offense/offense_play.cpp]: should this single defender have centre alignment when it is assigned (for example in crease_defender) ? do you know if we already handle that in crease_defender?

**Review by amrm3lm — COMMENTED** [review]
LGTM wondering if you were able to test that it works?

File: src/software/ai/hl/stp/play/offense/offense_play.cpp (hunk 2 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/offense/offense_play.cpp]: ternary statements are highly discouraged

**Review by jonathanlew — COMMENTED** [review]
I think it makes sense, just a nit

File: src/software/ai/hl/stp/play/offense/offense_play.cpp (hunk 2 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/offense/offense_play.cpp]: I think it's handled in the crease defense play

**Review by amrm3lm — DISMISSED** [review]
lgtm

File: src/software/ai/hl/stp/play/offense/offense_play.cpp [hunk 3]
```diff
@@ -26,20 +26,30 @@ void OffensePlay::getNextTactics(TacticCoroutine::push_type &yield, const World
...
+    PriorityTacticVector tactics_to_return;
+    unsigned int num_defenders = 2;
+    unsigned int num_enemy_robots =
+        static_cast<int>(play_update.world.enemyTeam().numRobots());
+
+    if (num_enemy_robots <= 3)
+    {
+        num_defenders = 1;
+        if (num_enemy_robots < 2)
+            num_defenders--;
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/play/offense/offense_play.cpp]: nit missing {}

File: src/software/ai/hl/stp/play/offense/offense_play.cpp (hunk 1 above)
Comment (akhilveeraghanta) [reply src/software/ai/hl/stp/play/offense/offense_play.cpp:41]: can you add this msg as a comment

**Review by akhilveeraghanta — COMMENTED** [review]
two minor nits, other than that lgtm!

File: src/software/ai/hl/stp/play/offense/offense_play.cpp (hunk 3 above)
Comment (itsarune) [inline src/software/ai/hl/stp/play/offense/offense_play.cpp]: doesn't this mean we just have 0 defender robots when there are less than two enemy robots, may as well just set it to 0 to be clearer

**Review by itsarune — APPROVED** [review]

**Review by akhilveeraghanta — APPROVED** [review]
lgtm!

**Pull request merged**